frame/transaction-payment/src/lib.rs

```diff
@@ -197,15 +232,30 @@ impl<T: Trait + Send + Sync> sp_std::fmt::Debug for ChargeTransactionPayment<T>
 }
 
 impl<T: Trait + Send + Sync> SignedExtension for ChargeTransactionPayment<T>
-	where BalanceOf<T>: Send + Sync
+where
+	BalanceOf<T>: Send + Sync,
 {
 	const IDENTIFIER: &'static str = "ChargeTransactionPayment";
 	type AccountId = T::AccountId;
 	type Call = T::Call;
 	type AdditionalSigned = ();
 	type DispatchInfo = DispatchInfo;
-	type Pre = ();
-	fn additional_signed(&self) -> sp_std::result::Result<(), TransactionValidityError> { Ok(()) }
+	type Pre = Option<PendingRefund<T::AccountId, NegativeImbalanceOf<T>>>;
+	fn additional_signed(&self) -> sp_std::result::Result<(), TransactionValidityError> {
+		Ok(())
+	}
+
+	fn pre_dispatch(
+		self,
+		who: &Self::AccountId,
+		_call: &Self::Call,
+		info: Self::DispatchInfo,
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		let tip = self.0;
+		let fee = Self::compute_fee(len as u32, info, tip);
+		Self::withdraw_fee(who, tip, fee).map_err(|()| InvalidTransaction::Payment.into())
```

Comment:
Let see if this is safe:

- If I panic within the code of the call itself, we have already charged the fee. That is good.
- But for example if the fee was meant to go to authors and treasury, we don't use the imbalance and it is just dropped, so we burn it. Is this desired? Or maybe it is addressed further down in the code. But we should def. discuss this scenario.

Reply:
@thiolliere actually, you are also changing this code so pinging you